contracts/upgradeability/Proxy.sol

```diff
@@ -13,13 +13,25 @@ contract Proxy {
   */
     function implementation() public view returns (address);
 
+    function setImplementation(address _newImplementation) external;
+
   /**
   * @dev Fallback function allowing to perform a delegatecall to the given implementation.
   * This function will return whatever the implementation call returns
   */
     function () payable public {
         address _impl = implementation();
         require(_impl != address(0));
+
+        address _innerImpl;
```

Comment:
the same proxy contract is used for validators contract. Could you clarify how these changes will affect it?

Reply:
Sorry, I didn't understand.
This is just to solve the nested proxy infinite loop problem.
If there is no nested proxy, meaning the implementation has no `implementation()` function, then the old proxy `delegatecall` will take place. I don't see where it could break stuff, but maybe I missed something.

Reply:
You are changing the contract that is used in several for the validator contract as well. How your changes will affect this contract?

Reply:
IMHO, no affect other functionality in other contracts, as long as you don't use an implementation with `implementation()` method in it, unless it is the nested proxy, as intended here.
But a good comprehensive audit should take place in ant case.

Reply:
is my understanding correct and we will increase gas usage of every call by two SSTORE command?

Reply:
Only if you use the double proxy. If not, no new sstore are added

Reply:
OK. Understood.